examples/firewall/firewall.c

```diff
@@ -0,0 +1,373 @@
+                                        RTE_LOG(INFO, APP, "Option -%c requires an argument.\n", optopt);
+                                if (optopt == 'f')
+                                        RTE_LOG(INFO, APP, "Option -%c requires an argument.\n", optopt);
+                                else if (isprint(optopt))
+                                        RTE_LOG(INFO, APP, "Unknown option `-%c'.\n", optopt);
+                                else
+                                        RTE_LOG(INFO, APP, "Unknown option character `\\x%x'.\n", optopt);
+                                return -1;
+                        default:
+                                usage(progname);
+                                return -1;
+                }
+        }
+
+        if (!dst_flag) {
+                RTE_LOG(INFO, APP, "Firewall NF requires a destination NF with the -d flag.\n");
+                return -1;
+        }
+        if (!debug) {
+                RTE_LOG(INFO, APP, "Running normal mode, use -b flag to enable debug mode\n");
+        }
+        if (!rules_init) {
+                RTE_LOG(INFO, APP, "Please specify a rules JSON file with -f FILE_NAME\n");
+                return -1;
+        }
+        return optind;
+}
+
+/*
+ * This function displays stats. It uses ANSI terminal codes to clear
+ * screen when called. It is called from a single non-master
+ * thread in the server process, when the process is run with more
+ * than one lcore enabled.
+ */
+static void
+do_stats_display(void) {
+        const char clr[] = {27, '[', '2', 'J', '\0'};
+        const char topLeft[] = {27, '[', '1', ';', '1', 'H', '\0'};
+
+        /* Clear screen and move to top left */
+        printf("%s%s", clr, topLeft);
+        printf("Packets Dropped: %lu\n", stats.pkt_drop);
+        printf("Packets not IPv4: %lu\n", stats.pkt_not_ipv4);
+        printf("Packets Accepted: %lu\n", stats.pkt_accept);
+        printf("Packets Total: %lu", stats.pkt_total);
+
+        printf("\n\n");
+}
+
+static int
+packet_handler(struct rte_mbuf *pkt, struct onvm_pkt_meta *meta, __attribute__((unused)) struct onvm_nf_info *nf_info) {
+        struct ipv4_hdr *ipv4_hdr;
+        static uint32_t counter = 0;
+        int ret;
+        uint32_t rule = 0;
+        uint32_t track_ip = 0;
+        uint32_t ip;
+        char ip_disp[16];
+
+        if (++counter == print_delay) {
+                do_stats_display();
+                counter = 0;
+        }
+
+        stats.pkt_total++;
+
+        if (!onvm_pkt_is_ipv4(pkt)) {
+                if (debug) RTE_LOG(INFO, APP, "Packet received not ipv4\n");
+                stats.pkt_not_ipv4++;
+                meta->action = ONVM_NF_ACTION_DROP;
+                return 0;
+        }
+
+        ipv4_hdr = onvm_pkt_ipv4_hdr(pkt);
+        ret = rte_lpm_lookup(lpm_tbl, rte_be_to_cpu_32(ipv4_hdr->src_addr), &rule);
+        sprintf(ip_disp, "%u.%u.%u.%u", (ip >> 24) & 0xFF, (ip >> 16) & 0xFF, (ip >> 8) & 0xFF, ip & 0xFF);
```

Comment:
This feels expensive it should only be done in debug move. Also might want to define a helper function smth like ip_to_string and put that into nflib as we use this often